subsys/usb/usb_device.c

```diff
@@ -1600,17 +1608,20 @@ int usb_enable(void)
 	/* Enable control EP */
 	ret = usb_dc_ep_enable(USB_CONTROL_OUT_EP0);
 	if (ret < 0) {
-		return ret;
+		goto usb_enable_error;
 	}
 
 	ret = usb_dc_ep_enable(USB_CONTROL_IN_EP0);
 	if (ret < 0) {
-		return ret;
+		goto usb_enable_error;
 	}
 
 	usb_dev.enabled = true;
```

Comment:
To be consistent with the documented behavior of `usb_enable()` this should be added, in case some of the intermediate functions returned a positive value:

    ret = 0;

Reply:
To be honest I dont get it. if failure, then return success?

Reply:
API documented behavior for `usb_enable()` is  _"0 on success, negative errno code on fail."_, so it is undefined for (ret > 0) :-)

But as we are here, should we cleanup and change this` if (ret < 0) `to `if (ret)`?
- `static int usb_vbus_set(bool on);` returns 0 on success, negative errno code on fail
- `int usb_dc_attach(void); `returns 0 on success, negative errno code on fail
- `int usb_dc_ep_configure(const struct usb_dc_ep_cfg_data * const cfg);` returns 0 on success, negative errno code on fail
- `int usb_dc_ep_set_callback(const u8_t ep, const usb_dc_ep_callback cb);` returns 0 on success, negative errno code on fail
- `static int composite_setup_ep_cb(void); `returns 0 on success, negative errno code on fail
- `int usb_dc_ep_enable(const u8_t ep);` returns 0 on success, negative errno code on fail

Reply:
> API documented behavior for `usb_enable()` is _"0 on success, negative errno code on fail."_, so it is undefined for (ret > 0) :-)

Exactly.  Failing to describe the complete value range of input and output parameters is pervasive problem in Zephyr.

> But as we are here, should we cleanup and change this`if (ret < 0)`to `if (ret)`?

Changing to `if (ret != 0)` would be acceptable as it satisfies the return guarantee--as long as none of those functions return positive.  Making a maintainer search for the documented (and implemented) behavior of all those subordinate functions to confirm the API  is not reasonable IMO, hence the change request.

Reply:
@emob-nordic Do you agree with "Changing to if (ret != 0) would be acceptable" ?

Reply:
Yes, now I get the point. I will update it in a minute.